### PR TITLE
gpu: sycl: shuffle: Enabling bwd direction

### PR DESCRIPTION
**Pull request by kala855**
# Description

The SYCL shuffle implementation does not support backward direction. This PR enables this feature.

### CODE REVIEW

File: src/gpu/generic/sycl/ref_shuffle.cpp [hunk 1]
```diff
@@ -43,18 +47,19 @@ status_t ref_shuffle_t::pd_t::init_conf() {
...
         conf_.HW = conf_.H * conf_.W;
         conf_.SP = conf_.D * conf_.HW;
     }
-    conf_.stat_md = xpu::sycl::md_t(src_md(0));
-    conf_.work_amount = memory_desc_wrapper(src_md()).nelems();
-    conf_.src_md = xpu::sycl::md_t(src_md(0));
-    conf_.dst_md = xpu::sycl::md_t(dst_md(0));
+    conf_.stat_md = xpu::sycl::md_t(src_data_md);
+    conf_.work_amount
+            = memory_desc_wrapper(is_fwd() ? src_md() : diff_src_md()).nelems();
```
Comment (t4c1) [inline src/gpu/generic/sycl/ref_shuffle.cpp:60]:
```suggestion
            = data_d.nelems();
```

File: src/gpu/generic/sycl/ref_shuffle.cpp [hunk 2]
```diff
@@ -84,7 +89,9 @@ status_t ref_shuffle_t::pd_t::init_conf() {
     const int wg_cnt = (t_work + wg_work - 1) / wg_work;
     conf_.nthr = wg_cnt * conf_.wg_size;
 
-    if (memory_desc_wrapper(src_md()).nelems() % conf_.block_size != 0)
+    if (memory_desc_wrapper(is_fwd() ? src_md() : diff_src_md()).nelems()
```
Comment (t4c1) [inline src/gpu/generic/sycl/ref_shuffle.cpp:92]:
```suggestion
    if (data_d.nelems()
```

Comment (kala855) [reply src/gpu/generic/sycl/ref_shuffle.cpp]: Suggested change accepted

**Review by t4c1 — APPROVED** [review]

File: src/gpu/generic/sycl/ref_shuffle.cpp [hunk 3]
```diff
@@ -28,7 +28,11 @@ using namespace impl::utils;
 
 status_t ref_shuffle_t::pd_t::init_conf() {
     conf_ = sycl_shuffle_conf_t();
-    const memory_desc_wrapper data_d(src_md(0));
+
+    auto src_data_md = is_fwd() ? src_md(0) : diff_src_md(0);
```
Comment (dzarukin) [inline src/gpu/generic/sycl/ref_shuffle.cpp:32]: There are `invariant_src_md()` and `invariant_dst_md()` methods in `primitive_desc.hpp` that would handle it for the user.

File: src/gpu/generic/sycl/ref_shuffle.hpp [hunk 4]
```diff
@@ -42,18 +42,20 @@ struct ref_shuffle_t : public gpu::generic::sycl::primitive_t {
...
-                            || src_md()->data_type == f32)
-                    && (src_md(0)->format_desc.blocking.inner_nblks == 0)
-                    && attr()->has_default_values()
+            const bool ok = src_data_md->data_type == dst_data_md->data_type
+                    && (src_data_md->data_type == bf16
+                            || src_data_md->data_type == f16
+                            || src_data_md->data_type == f32
+                            || (is_fwd() && src_data_md->data_type == s32)
+                            || (is_fwd() && src_data_md->data_type == s8)
+                            || (is_fwd() && src_data_md->data_type == u8))
```
Comment (dzarukin) [inline src/gpu/generic/sycl/ref_shuffle.hpp:54]: There's `utils::one_of(src_data_type, dt0, dt1, ...)` function for such enumeration.

File: src/gpu/generic/sycl/ref_shuffle.cpp (hunk 3 above)
Comment (kala855) [reply src/gpu/generic/sycl/ref_shuffle.cpp]: Suggestion made. Thank you.

File: src/gpu/generic/sycl/ref_shuffle.hpp (hunk 4 above)
Comment (kala855) [reply src/gpu/generic/sycl/ref_shuffle.hpp]: Suggestions included thanks.

**Review by dzarukin — APPROVED** [review]

**Pull request merged**